### DISCUSSION TITLE
Searching with vector A, different than searching with vector B

### DISCUSSION THREAD

**Discussion by sickerin** [milvus-io/milvus#27613] [Q&A and General discussion]
Hi, I have a question regarding whether the search with vector A and vector B should have different distance scores for each other. Is there an info page that addresses this? Whether the distance between A and B should quite similar to the distance between B and A.

The images are face embeddings inserted into milvus. 

Image A: top results
- Image A (score 0.8742654323577881)
- Image B (score 0.561960279941558)

Image B: top results
- Image B (score 0.906684935092926)
- Image A (score 0.6493653059005737)

```python
            param = {
                'collection_name': collection_name,
                'query_records'  : embedding_vec,
                'top_k'          :  10, 
                'params'         : 30,
            }
            status, results = milvus_client.search(**param)
```

```python
  DIM = 512
    METRIC_TYPE = "IP"
    INDEX_FILE_SIZE = 1024
    INDEX_CONFIG = {
        "index_type": "HNSW",
        "index_param": {
            'M': 48,
            'efConstruction': 300,
        }
    }
```

Version:  pymilvus==1.1.1

**Comment by yhmo** [comment]
With metric type="IP", all the vectors should be normalized(length=1.0).
If all the vectors are normalized, you search image_A, the top results will be:
image A (score 1.0)
image B (score <1.0)


Refer to this document: https://milvus.io/docs/metric.md#Similarity-Metrics

**Comment by sickerin** [comment]
I see, thanks. Yes, I was surprised that I didn't get sth close to 1.0 or exactly 1.0 since it's an inner product of itself. I have normalized using the following.

`embedding = sklearn.preprocessing.normalize(embedding).flatten()`

Is the other part true? If I search image_A, and I get a score (0.561) for image_B, should I expect to get a score (0.561) for image_A when I search with image_B? Given that it's an inner product? Or does milvus do something during ANN where the vectors maybe slightly different depending on what is searched?

Currently, there's quite a bit of difference.

A search B -> Image B (score 0.561960279941558)
B search A -> Image A (score 0.6493653059005737)

If it shouldn't be quite different, then there's probably a bug in my code or sth somewhere. Even if I let's say add A last year, and I added B this year, and in between these two instances, 100,000 other vectors have been added.

**Reply by filip-halt** [reply]
That's interesting. The dot product (IP) should be commutative, so a.b = b.a , can I see your code block by chance? Were there any floating point conversions being done?

**Comment by yhmo** [comment]
I write a script to try your case, the result is correct. When I search A or B, the product value is equal.

```
import random
import json
from sklearn import preprocessing
import numpy as np

from pymilvus import (
    connections,
    FieldSchema, CollectionSchema, DataType,
    Collection,
    utility,
)


HOST = '127.0.0.1'
PORT = '19530'

# Const names
COLLECTION_NAME = 'demo'
ID_FIELD_NAME = 'id'
VECTOR_FIELD_NAME = 'vector'

DIM = 8

def create_collection():
    if utility.has_collection(COLLECTION_NAME):
        utility.drop_collection(COLLECTION_NAME)

    field1 = FieldSchema(name=ID_FIELD_NAME, dtype=DataType.INT64, is_primary=True, auto_id=False)
    field2 = FieldSchema(name=VECTOR_FIELD_NAME, dtype=DataType.FLOAT_VECTOR, dim=DIM)
    schema = CollectionSchema(fields=[field1, field2])
    collection = Collection(name=COLLECTION_NAME, schema=schema)
    print("\ncollection created:", COLLECTION_NAME)
    return collection

def gen_embedding():
    embedding = np.array([random.random() for _ in range(DIM)])
    print("original:", embedding)
    embedding = preprocessing.normalize([embedding]).flatten()
    print("normalized", embedding)
    return embedding

if __name__ == '__main__':
    connections.connect(
        alias="default",
        host=HOST,
        port=PORT,
    )
    print(utility.get_server_version())

    create_collection()

    collection = Collection(COLLECTION_NAME)
    index = {
        "index_type": "FLAT",
        "metric_type": "IP",
        "params": {},
    }
    collection.create_index(VECTOR_FIELD_NAME, index)
    collection.load()

    print("generate embedding A:")
    embedding_A = gen_embedding()
    print("generate embedding B:")
    embedding_B = gen_embedding()

    data = [
        [1, 2],
        [embedding_A, embedding_B],
    ]
    collection.insert(data=data)
    collection.flush()
    print("row count:", collection.num_entities)

    search_param = {
        "data": [embedding_A],
        "anns_field": VECTOR_FIELD_NAME,
        "param": {"metric_type": "IP", "params": {}},
        "limit": 10,
    }
    results = collection.search(**search_param)
    for i, result in enumerate(results):
        print("\nSearch result for A: ".format(i))
        for j, res in enumerate(result):
            print("Top {}: {}".format(j, res))

    search_param = {
        "data": [embedding_B],
        "anns_field": VECTOR_FIELD_NAME,
        "param": {"metric_type": "IP", "params": {}},
        "limit": 10,
    }
    results = collection.search(**search_param)
    for i, result in enumerate(results):
        print("\nSearch result for B: ".format(i))
        for j, res in enumerate(result):
            print("Top {}: {}".format(j, res))
```

Output:
```
v2.3.1

collection created: demo
generate embedding A:
original: [0.99498128 0.06098374 0.13677175 0.63736757 0.00377611 0.56723046
 0.50735935 0.64129052]
normalized [0.64104584 0.03929056 0.08811921 0.41064273 0.00243287 0.36545484
 0.32688112 0.4131702 ]
generate embedding B:
original: [0.25510177 0.73245205 0.73326467 0.29604828 0.47193319 0.40091759
 0.58306472 0.30982752]
normalized [0.1783319  0.51202924 0.51259731 0.20695604 0.32991046 0.28026617
 0.40759826 0.21658858]
row count: 2

Search result for A: 
Top 0: id: 1, distance: 0.9999998807907104, entity: {}
Top 1: id: 2, distance: 0.5905428528785706, entity: {}

Search result for B: 
Top 0: id: 2, distance: 1.0, entity: {}
Top 1: id: 1, distance: 0.5905428528785706, entity: {}

```

**Reply by sickerin** [reply]
Thanks a lot, I'll try to investigate further!
